### PR TITLE
SCC: adding datasources for automated WP connection  and  Profile attachments

### LINKED ISSUE #5185: SCC Profile Data Lookup By Name
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

I want to be able to lookup an scc profile by name instead of id. Our terraform code creates an scc profile, and on reapply we will need to lookup the profile again by name to check for existing controls and values, but we do not have access to the profile id at that time and therefore cannot provide existing values which results in terraform removing existing controls ect since they weren't provided.

Basically this [datasource](https://registry.terraform.io/providers/IBM-Cloud/ibm/latest/docs/data-sources/scc_profile#controls) should also accept a profile name and not have to provide an id.

### New or Affected Resource(s)

* ibm_scc_profile

### Potential Terraform Configuration

````
data "ibm_scc_profile" "scc_profile" {
    instance_id = "00000000-1111-2222-3333-444444444444"
    profile_name = "profile_1"
}
````
### References

Docs: https://registry.terraform.io/providers/IBM-Cloud/ibm/latest/docs/data-sources/scc_profile#controls
internal issue: https://github.ibm.com/project-fortress/pm/issues/15403

* #0000

### LINKED ISSUE #5186: SCC Control Library Data Lookup By Name
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

I want to be able to lookup an scc control library by name instead of id. Our terraform code creates an scc control library, and on reapply we will need to lookup the control library again by name to check for existing controls and values, but we do not have access to the `control_library_id` at that time and therefore cannot provide existing values which results in terraform removing existing controls ect. since they weren't provided.

Basically this [datasource](https://registry.terraform.io/providers/IBM-Cloud/ibm/latest/docs/data-sources/scc_profile#controls) should also accept a control library name and not have to provide an id.

### New or Affected Resource(s)

* ibm_scc_control_library

### Potential Terraform Configuration

````
data "ibm_scc_control_library" "scc_control_library" {
    instance_id = "00000000-1111-2222-3333-444444444444"
    control_library_name = "control_name"
}
````
### References

Docs: https://registry.terraform.io/providers/IBM-Cloud/ibm/latest/docs/data-sources/scc_control_library
internal issue: https://github.ibm.com/project-fortress/pm/issues/15403

* #0000

### LINKED ISSUE #5208: Add the ability to do data lookup on ibm_scc_provider_type using name
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

The `ibm_scc_provider_type_instance` resource requires you to pass a value for `provider_type_id`, however the [ibm_scc_provider_type](https://registry.terraform.io/providers/ibm-cloud/ibm/latest/docs/data-sources/scc_provider_type) data lookup only allows you to do a lookup by ID. 
This makes it impossible to use `ibm_scc_provider_type_instance` without already knowing the ID.

### New or Affected Resource(s)

* ibm_scc_provider_type

### Potential Terraform Configuration

```hcl
data "ibm_scc_provider_type" "scc_provider_type" {
  name  = "Security and Compliance Workload Protection"
  region = var.region
}
```

### References

* #0000

### PR DESCRIPTION
**Pull request by tyao117**
Updates:
* expose a new datasource `ibm_scc_control_libraries` that returns a list of control_libraries available to an instance.
  * Users can now use the datasource to help create a profile resource given from the datasource. 
* expose a new datasource `ibm_scc_profiles` that returns a list of profiles available to an instance
  * Users can now use the datasource to help create a profile_attachment resource in relation to the data given from the datasource 
* expose a new datasource `ibm_scc_provider_types` that returns a list of provider available to an instance 
  * Users can use the datasource to find the ibm_scc_provider_type.id  needed to create a resource of  scc_provider_type_instance. 
* fixes the import of documentation of scc
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #5186 
Closes #5185
Closes #5208

Output from acceptance testing:

```shell
$ make testacc TEST=./ibm/service/scc       
=== RUN   TestAccIbmSccControlLibrariesDataSourceBasic
--- PASS: TestAccIbmSccControlLibrariesDataSourceBasic (51.25s)
=== RUN   TestAccIbmSccControlLibrariesDataSourceAllArgs
--- PASS: TestAccIbmSccControlLibrariesDataSourceAllArgs (41.44s)
=== RUN   TestAccIbmSccControlLibraryDataSourceBasic
--- PASS: TestAccIbmSccControlLibraryDataSourceBasic (56.03s)
=== RUN   TestAccIbmSccControlLibraryDataSourceAllArgs
--- PASS: TestAccIbmSccControlLibraryDataSourceAllArgs (55.85s)
=== RUN   TestAccIbmSccInstanceSettingsDataSourceBasic
--- PASS: TestAccIbmSccInstanceSettingsDataSourceBasic (42.82s)
=== RUN   TestAccIbmSccLatestReportsDataSourceBasic
--- PASS: TestAccIbmSccLatestReportsDataSourceBasic (44.97s)
=== RUN   TestAccIbmSccProfileAttachmentDataSourceBasic
--- PASS: TestAccIbmSccProfileAttachmentDataSourceBasic (66.66s)
=== RUN   TestAccIbmSccProfileAttachmentDataSourceAllArgs
--- PASS: TestAccIbmSccProfileAttachmentDataSourceAllArgs (64.63s)
=== RUN   TestAccIbmSccProfileDataSourceBasic
--- PASS: TestAccIbmSccProfileDataSourceBasic (65.19s)
=== RUN   TestAccIbmSccProfileDataSourceAllArgs
--- PASS: TestAccIbmSccProfileDataSourceAllArgs (58.10s)
=== RUN   TestAccIbmSccProfilesDataSourceBasic
--- PASS: TestAccIbmSccProfilesDataSourceBasic (44.98s)
=== RUN   TestAccIbmSccProfilesDataSourceAllArgs
--- PASS: TestAccIbmSccProfilesDataSourceAllArgs (48.66s)
=== RUN   TestAccIbmSccProviderTypeCollectionDataSourceBasic
--- PASS: TestAccIbmSccProviderTypeCollectionDataSourceBasic (40.12s)
=== RUN   TestAccIbmSccProviderTypeInstanceDataSourceBasic
--- PASS: TestAccIbmSccProviderTypeInstanceDataSourceBasic (51.32s)
=== RUN   TestAccIbmSccProviderTypeInstanceDataSourceAllArgs
--- PASS: TestAccIbmSccProviderTypeInstanceDataSourceAllArgs (51.91s)
=== RUN   TestAccIbmSccProviderTypeDataSourceBasic
--- PASS: TestAccIbmSccProviderTypeDataSourceBasic (39.37s)
=== RUN   TestAccIbmSccProviderTypesDataSourceBasic
--- PASS: TestAccIbmSccProviderTypesDataSourceBasic (39.61s)
=== RUN   TestAccIbmSccReportControlsDataSourceBasic
--- PASS: TestAccIbmSccReportControlsDataSourceBasic (42.73s)
=== RUN   TestAccIbmSccReportEvaluationsDataSourceBasic
--- PASS: TestAccIbmSccReportEvaluationsDataSourceBasic (42.10s)
=== RUN   TestAccIbmSccReportResourcesDataSourceBasic
--- PASS: TestAccIbmSccReportResourcesDataSourceBasic (39.08s)
=== RUN   TestAccIbmSccReportRuleDataSourceBasic
--- PASS: TestAccIbmSccReportRuleDataSourceBasic (39.16s)
=== RUN   TestAccIbmSccReportSummaryDataSourceBasic
--- PASS: TestAccIbmSccReportSummaryDataSourceBasic (44.01s)
=== RUN   TestAccIbmSccReportTagsDataSourceBasic
--- PASS: TestAccIbmSccReportTagsDataSourceBasic (38.63s)
=== RUN   TestAccIbmSccReportDataSourceBasic
--- PASS: TestAccIbmSccReportDataSourceBasic (38.64s)
=== RUN   TestAccIbmSccReportViolationDriftDataSourceBasic
--- PASS: TestAccIbmSccReportViolationDriftDataSourceBasic (39.60s)
=== RUN   TestAccIbmSccRuleDataSourceBasic
--- PASS: TestAccIbmSccRuleDataSourceBasic (56.10s)
=== RUN   TestAccIbmSccRuleDataSourceAllArgs
--- PASS: TestAccIbmSccRuleDataSourceAllArgs (55.48s)
=== RUN   TestAccIbmSccControlLibraryBasic
--- PASS: TestAccIbmSccControlLibraryBasic (87.26s)
=== RUN   TestAccIbmSccControlLibraryAllArgs
--- PASS: TestAccIbmSccControlLibraryAllArgs (105.08s)
=== RUN   TestAccIbmSccInstanceSettingsBasic
--- PASS: TestAccIbmSccInstanceSettingsBasic (56.88s)
=== RUN   TestAccIbmSccInstanceSettingsAllArgs
--- PASS: TestAccIbmSccInstanceSettingsAllArgs (111.14s)
=== RUN   TestAccIbmSccProfileAttachmentBasic
--- PASS: TestAccIbmSccProfileAttachmentBasic (63.77s)
=== RUN   TestAccIbmSccProfileAttachmentAllArgs
--- PASS: TestAccIbmSccProfileAttachmentAllArgs (104.06s)
=== RUN   TestAccIbmSccProfileBasic
--- PASS: TestAccIbmSccProfileBasic (92.74s)
=== RUN   TestAccIbmSccProfileAllArgs
--- PASS: TestAccIbmSccProfileAllArgs (97.96s)
=== RUN   TestAccIbmSccProviderTypeInstanceBasic
--- PASS: TestAccIbmSccProviderTypeInstanceBasic (89.87s)
=== RUN   TestAccIbmSccProviderTypeInstanceAllArgs
--- PASS: TestAccIbmSccProviderTypeInstanceAllArgs (98.29s)
=== RUN   TestAccIbmSccRuleBasic
--- PASS: TestAccIbmSccRuleBasic (104.38s)
=== RUN   TestAccIbmSccRuleAllArgs
--- PASS: TestAccIbmSccRuleAllArgs (99.30s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/scc     2409.749s
```

### CODE REVIEW

**Comment by jor2** [comment]
could we get controls too pls https://github.com/IBM-Cloud/terraform-provider-ibm/blob/5dfec705149c51e91b0a8d956a7f9eea891e6ee0/ibm/service/scc/data_source_ibm_scc_profile.go#L104

**Comment by jor2** [comment]
could we get controls in the schema too pls, like here https://github.com/IBM-Cloud/terraform-provider-ibm/blob/5dfec705149c51e91b0a8d956a7f9eea891e6ee0/ibm/service/scc/data_source_ibm_scc_profile.go#L104.

**Comment by tyao117** [comment]
> could we get controls in the schema too pls, like here
> 
> https://github.com/IBM-Cloud/terraform-provider-ibm/blob/5dfec705149c51e91b0a8d956a7f9eea891e6ee0/ibm/service/scc/data_source_ibm_scc_profile.go#L104
> 
> .

I can implement this but it will slow down this datasource **significantly**. I would need to query every single profile for their controls. 

here is a run without the controls included:
```shell
make testacc TEST=./ibm/service/scc  TESTARGS="-run TestAccIbmSccProfilesData*"
=== RUN   TestAccIbmSccProfilesDataSourceBasic
--- PASS: TestAccIbmSccProfilesDataSourceBasic (50.92s)
PASS
ok      github.com/IBM-Cloud/terraform-provider-ibm/ibm/service/scc     51.344s
```
my run with the control doesn't even finish in a reasonable time. After running it the same `make command`, it is still running  with my cpu time
image

i would advise using the datasource once you find the profile_id as an alternative.

**Comment by tyao117** [comment]
needing to push documentation out for the new datasources

**Comment by tyao117** [comment]
need to add datasource_ibm_control_libraries

**Comment by tyao117** [comment]
datasource documentation added.

File: ibm/service/scc/resource_ibm_scc_control_library.go [hunk 1]
```diff
@@ -876,6 +887,17 @@ func resourceIbmSccControlLibraryControlsInControlLibToMap(model *securityandcom
 	return modelMap, nil
 }
 
+// using the assessment_id for comparison
+func compareAssessmentSetFunc(v interface{}) int {
+	m := v.(map[string]interface{})
```
Comment (hkantare) [inline ibm/service/scc/resource_ibm_scc_control_library.go:892]: Can you add some checks to check len of map and element of map not nil to eliminate any nil pointer exceptions

File: ibm/service/scc/resource_ibm_scc_profile_attachment.go [hunk 2]
```diff
@@ -321,6 +324,16 @@ func resourceIbmSccProfileAttachmentCreate(context context.Context, d *schema.Re
 	return resourceIbmSccProfileAttachmentRead(context, d, meta)
 }
 
+func cmpAttachParamSetFunc(v interface{}) int {
```
Comment (hkantare) [inline ibm/service/scc/resource_ibm_scc_profile_attachment.go:327]: same above commands holds here also

File: ibm/service/scc/resource_ibm_scc_profile_attachment.go [hunk 3]
```diff
@@ -321,6 +324,16 @@ func resourceIbmSccProfileAttachmentCreate(context context.Context, d *schema.Re
...
+	m := v.(map[string]interface{})
+	id := (m["assessment_id"]).(*string)
+	assId := (*id)[5:18]
+	var i big.Int
+	i.SetString(strings.Replace(assId, "-", "", 4), 16)
+	val, _ := strconv.Atoi(i.String())
```
Comment (hkantare) [inline ibm/service/scc/resource_ibm_scc_profile_attachment.go:333]: don't supress errors log/throw back the error

**Comment by tyao117** [comment]
i have implemented the checks that you have asked for.

File: ibm/service/scc/resource_ibm_scc_control_library.go (hunk 1 above)
Comment (hkantare) [reply ibm/service/scc/resource_ibm_scc_control_library.go:895]: In schema assessment_id is defined as optional if user define 
assessments = {
assessment_method = "XXXXX"
}

**Review by hkantare — APPROVED** [review]

**Pull request merged**